src/psyclone/psyir/symbols/symbol.py

```diff
@@ -255,6 +255,32 @@ def copy(self):
         return type(self)(self.name, visibility=self.visibility,
                           interface=self.interface)
 
+    def specialise(self, subclass):
+        '''Specialise this symbol so that it becomes an instance of the class
+        provided in the subclass argument. This allows this instance
+        to become a subclass without any references to it becoming
+        invalid.
+
+        :param subclass: the class that this symbol will become.
+        :type subclass: `type` (as `type` is the type of all classes)
```

Comment:
Unless I'm missing something I don't think type-hinting would help this particular problem as it would not help us describe and render the interface in Sphinx. I do like the idea of type hinting though, although I'm not sure how it fits in with interface documentation. Perhaps you could bring it up at the next telco?

What I've done is describe the class, rather than the type of the class. It is not strictly correct - I don't think - but is probably what people would expect to see - perhaps!

Reply:
@rupertford I didn't explain it well. I was not suggesting type hinting yet (I don't think python2 would like it), I was suggesting using elements of its syntax for the ":type" fields of the docstring as it has a proper definition of each case and I think Sphinx is implemented to recognize this syntax. But it doesn't really matter for this issue, maybe a future discussion.

Anyway, I do agree with you that the plain class is not strictly correct (as it not the instance we are passing here).
Since this is the only issue remaining I will change it before merging from the current:
``` :type subclass: sub-class of :py:class:`psyclone.psyir.symbols.Symbol```
to 
``` :type subclass: type of sub-class of :py:class:`psyclone.psyir.symbols.Symbol```
unless you disagree.

Reply:
Updated.